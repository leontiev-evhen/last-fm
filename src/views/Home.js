import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { ListItem, ErrorBoundary } from '../components';
import { fetchTracks } from '../actions';
import ContainerHOC from '../hoc/ContainerHOC';

const DivContainerHOC = ContainerHOC(({ children }) => children);

export class Home extends Component {
	componentDidMount() {
		if (this.props.tracks.length === 0) this.props.fetchTopTracks();
	}

	render() {
		const { tracks, isFetching, error } = this.props;
		return (
			<div className="row">
				<DivContainerHOC isFetching={isFetching} error={error}>
					{tracks.map((track, index) => (
						<ErrorBoundary key={index}>
              <ListItem track={track} />
            </ErrorBoundary>
					))}
				</DivContainerHOC>
			</div>
		);
	}
}

Home.propTypes = {
	tracks: PropTypes.array.isRequired,
	isFetching: PropTypes.bool.isRequired,
	error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
	fetchTopTracks: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
	tracks: state.tracks.tracks,
	isFetching: state.tracks.isFetching,
	error: state.tracks.error,
});

const mapDispatchToProps = dispatch => ({
	fetchTopTracks: () => {
		dispatch(fetchTracks());
	},
});

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(Home);
