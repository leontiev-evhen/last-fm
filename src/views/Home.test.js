import React from 'react';
import { shallow } from 'enzyme';
import { Home } from './Home';


describe('Home view', () => {
  const props = {
    tracks: [],
    isFetching: false,
    error: '',
    fetchTopTracks: () => {},
  };

  it('Home component componentDidMount', () => {
    jest.spyOn(Home.prototype, 'componentDidMount');
    shallow(<Home {...props} />);
    expect(Home.prototype.componentDidMount).toHaveBeenCalled();
  });
});
