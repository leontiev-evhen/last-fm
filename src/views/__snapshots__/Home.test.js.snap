// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home view render correctly Home component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    error=""
    fetchTopTracks={[Function]}
    isFetching={false}
    tracks={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <containerHOC
        error=""
        isFetching={false}
      />,
      "className": "row",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [],
        "error": "",
        "isFetching": false,
      },
      "ref": null,
      "rendered": Array [],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <containerHOC
          error=""
          isFetching={false}
        />,
        "className": "row",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [],
          "error": "",
          "isFetching": false,
        },
        "ref": null,
        "rendered": Array [],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
